<a href="https://colab.research.google.com/github/konradcodes1/university-majors-eda-analysis/blob/main/majors_eda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploratory Data Analysis

Importowanie bibliotek

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Zaczynam od wczytania danych i wyświetlenia pierwszych 10 wierszy, aby wstępnie ocenić strukturę zbioru danych.

In [ ]:
df = pd.read_csv('college_major_roi.csv')
df.head(10)

,grad_id,major,major_category,institution_tier,institution_selectivity_pctile,gpa,had_internship,completed_on_time,region,net_cost_usd,debt_usd,hs_baseline_10yr_usd,added_earnings_10yr_usd,earnings_10yr_usd,net_roi_usd,roi_pct,positive_roi,high_roi
0,G000000,biology,STEM,mid_tier,52,2.58,0,0,south,144400,0,411700,101000,512700,-43400,-30.0,0,0
1,G000001,marketing,business,mid_tier,61,2.58,0,1,northeast,57400,0,356900,330600,687500,273200,476.1,1,0
2,G000002,electrical_engineering,STEM,top_public,78,3.14,1,1,northeast,65100,39600,339600,655100,994700,590000,906.4,1,1
3,G000003,marketing,business,top_public,77,3.04,0,1,midwest,105800,81300,399600,197500,597100,91700,86.8,1,0
4,G000004,marketing,business,ivy_elite,100,3.81,0,0,south,219200,93500,335700,222900,558600,3800,1.7,1,0
5,G000005,business,business,regional,37,3.91,1,0,northeast,79000,0,413800,466800,880600,387900,491.2,1,0
6,G000006,economics,business,regional,44,2.32,1,1,midwest,39500,0,388800,405200,794100,365800,926.3,1,0
7,G000007,social_work,social_science,mid_tier,59,4.00,1,1,midwest,59300,43500,350300,162400,512600,103100,174.0,1,0
8,G000008,mechanical_engineering,STEM,regional,45,3.61,0,1,northeast,62200,0,314300,344500,658800,282300,454.2,1,0
9,G000009,finance,business,top_public,79,2.46,1,0,south,109000,107500,307200,438000,745200,329000,301.8,1,0


Sprawdzam rozmiar zbioru, typy danych, liczbę brakujących wartości oraz podstawowe statystyki opisowe. Pozwoli to szybko wychwycić potencjalne problemy z jakością danych i lepiej zrozumieć strukturę datasetu.


In [ ]:
df.describe()

,institution_selectivity_pctile,gpa,had_internship,completed_on_time,net_cost_usd,debt_usd,hs_baseline_10yr_usd,added_earnings_10yr_usd,earnings_10yr_usd,net_roi_usd,roi_pct,positive_roi,high_roi
count,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,3.000000e+04,3.000000e+04,3.000000e+04,30000.000000,30000.000000,30000.00000
mean,55.086067,3.093337,0.436067,0.682367,88621.496667,38498.063333,359893.416667,3.554752e+05,7.153680e+05,2.668537e+05,346.153987,0.835167,0.25000
std,18.593708,0.439060,0.495904,0.465564,33576.903160,37801.748200,39923.943255,2.948109e+05,2.977847e+05,2.949958e+05,398.666879,0.371036,0.43302
min,16.000000,1.500000,0.000000,0.000000,28500.000000,0.000000,250000.000000,1.000000e+04,2.723000e+05,-3.052000e+05,-92.100000,0.000000,0.00000
25%,42.000000,2.790000,0.000000,0.000000,65500.000000,0.000000,333100.000000,1.179000e+05,4.835000e+05,3.280000e+04,39.700000,1.000000,0.00000
50%,55.000000,3.100000,0.000000,1.000000,81000.000000,38700.000000,359700.000000,2.836500e+05,6.458000e+05,1.981000e+05,230.900000,1.000000,0.00000
75%,68.000000,3.400000,1.000000,1.000000,103600.000000,62300.000000,387300.000000,5.093000e+05,8.717000e+05,4.235250e+05,537.400000,1.000000,0.25000
max,102.000000,4.000000,1.000000,1.000000,352900.000000,331700.000000,480000.000000,2.894900e+06,3.245500e+06,2.755100e+06,3618.900000,1.000000,1.00000


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   grad_id                         30000 non-null  object 
 1   major                           30000 non-null  object 
 2   major_category                  30000 non-null  object 
 3   institution_tier                30000 non-null  object 
 4   institution_selectivity_pctile  30000 non-null  int64  
 5   gpa                             30000 non-null  float64
 6   had_internship                  30000 non-null  int64  
 7   completed_on_time               30000 non-null  int64  
 8   region                          30000 non-null  object 
 9   net_cost_usd                    30000 non-null  int64  
 10  debt_usd                        30000 non-null  int64  
 11  hs_baseline_10yr_usd            30000 non-null  int64  
 12  added_earnings_10yr_usd         

Zbiór zawiera 30 000 wierszy i 18 kolumn. Wszystkie kolumny mają odpowiednie typy danych, dlatego na tym etapie nie są potrzebne żadne konwersje.

In [ ]:
binary_cols = [
    "had_internship",
    "completed_on_time",
    "positive_roi",
    "high_roi"
]

for col in binary_cols:
    print(df[col].value_counts())
    print()

had_internship
0    16918
1    13082
Name: count, dtype: int64

completed_on_time
1    20471
0     9529
Name: count, dtype: int64

positive_roi
1    25055
0     4945
Name: count, dtype: int64

high_roi
0    22500
1     7500
Name: count, dtype: int64



In [ ]:
df[df["institution_selectivity_pctile"] > 100]

,grad_id,major,major_category,institution_tier,institution_selectivity_pctile,gpa,had_internship,completed_on_time,region,net_cost_usd,debt_usd,hs_baseline_10yr_usd,added_earnings_10yr_usd,earnings_10yr_usd,net_roi_usd,roi_pct,positive_roi,high_roi
96,G000096,electrical_engineering,STEM,ivy_elite,102,2.67,0,0,west,227500,170600,374000,1476800,1850800,1249300,549.2,1,1
192,G000192,fine_arts,arts,ivy_elite,101,3.11,0,0,south,173000,0,295500,76600,372100,-96400,-55.7,0,0
307,G000307,finance,business,ivy_elite,101,3.15,0,1,west,154100,124300,358600,386000,744600,231900,150.4,1,0
415,G000415,marketing,business,ivy_elite,101,2.04,0,1,south,92000,0,362100,331300,693400,239300,260.1,1,0
470,G000470,biology,STEM,ivy_elite,102,3.47,1,0,south,160000,125700,264400,206000,470500,46100,28.8,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29473,G029473,social_work,social_science,ivy_elite,101,3.19,1,1,west,205900,194600,445100,113600,558800,-92200,-44.8,0,0
29730,G029730,chemical_engineering,STEM,ivy_elite,102,2.85,0,1,midwest,100600,0,357400,919600,1277000,819000,813.9,1,1
29754,G029754,education,education,ivy_elite,101,3.31,0,1,south,171000,81000,399300,99300,498600,-71700,-42.0,0,0
29800,G029800,electrical_engineering,STEM,ivy_elite,101,2.84,1,0,west,189700,0,348600,867700,1216300,678000,357.3,1,1


In [ ]:
invalid_pctile = df[df["institution_selectivity_pctile"] > 100]

invalid_pctile["institution_selectivity_pctile"].value_counts()

,count
institution_selectivity_pctile,
101,158
102,130


In [ ]:
invalid_pctile["institution_tier"].value_counts()

,count
institution_tier,
ivy_elite,288


In [ ]:
len(invalid_pctile) / len(df) * 100

0.96

Wartości powyżej 100 występują wyłącznie dla instytucji oznaczonych jako ivy_elite. Obejmują 288 obserwacji, czyli 0,96% zbioru, i przyjmują wyłącznie wartości 101 lub 102. Sugeruje to, że przekroczenie standardowego zakresu percentyla może wynikać ze sposobu konstrukcji zmiennej, a nie z przypadkowych błędów danych.

In [ ]:
df.groupby("institution_tier")["institution_selectivity_pctile"].agg(
    ["min", "max", "median"]
)

,min,max,median
institution_tier,,,
for_profit,16,28,22.0
ivy_elite,90,102,96.0
mid_tier,52,64,58.0
regional,36,48,42.0
top_public,68,80,74.0


In [ ]:
df[df["institution_tier"] == "ivy_elite"][
    "institution_selectivity_pctile"
].value_counts().sort_index()

,count
institution_selectivity_pctile,
90,137
91,124
92,137
93,124
94,129
95,127
96,136
97,147
98,131


Analiza wartości institution_selectivity_pctile według poziomu instytucji wykazała wyraźnie rozdzielone zakresy dla poszczególnych kategorii. Wartości 101–102 występują w ramach spójnego zakresu 90–102 dla ivy_elite, dlatego nie zostały uznane za błędne obserwacje. Zmienna prawdopodobnie pełni funkcję syntetycznego wskaźnika selektywności, mimo nazwy sugerującej klasyczny percentyl.

In [ ]:
roi_by_category = (
    df.groupby("major_category")
      .agg(
          count=("net_roi_usd", "size"),
          median_roi=("net_roi_usd", "median"),
          mean_roi=("net_roi_usd", "mean"),
          negative_roi_share=("net_roi_usd", lambda x: (x < 0).mean())
      )
      .sort_values("negative_roi_share", ascending=False)
)

roi_by_category

,count,median_roi,mean_roi,negative_roi_share
major_category,,,,
arts,1395,-35500.0,-39821.146953,0.888172
humanities,1410,-11200.0,-13321.205674,0.623404
education,1352,8750.0,11547.855030,0.417899
social_science,5989,22800.0,31064.017365,0.330272
business,8361,291700.0,346099.366105,0.015668
STEM,9904,393350.0,446999.798061,0.013126
health,1589,318500.0,350828.823159,0.008181


Analiza według kategorii kierunków pokazuje wyraźne różnice w wartości net_roi_usd. Kategorie arts i humanities wyróżniają się ujemną zarówno średnią, jak i medianą ROI, a większość obserwacji w tych grupach osiąga wynik poniżej zera. W przypadku social_science ujemne ROI występuje relatywnie często, jednak mediana i średnia pozostają dodatnie, co sugeruje, że wcześniejszy zaobserwowany przypadek nie jest reprezentatywny dla całej kategorii.

In [ ]:
roi_by_major = (
    df.groupby("major")
      .agg(
          count=("net_roi_usd", "size"),
          median_roi=("net_roi_usd", "median"),
          negative_roi_share=("net_roi_usd", lambda x: (x < 0).mean())
      )
      .sort_values("negative_roi_share", ascending=False)
)

roi_by_major.head(15)

,count,median_roi,negative_roi_share
major,,,
fine_arts,1395,-35500.0,0.888172
humanities,1410,-11200.0,0.623404
psychology,1429,1800.0,0.481456
social_work,1396,5050.0,0.447708
education,1352,8750.0,0.417899
criminal_justice,1376,33800.0,0.260901
communications,1788,60500.0,0.171141
biology,1735,119300.0,0.065706
marketing,1703,129300.0,0.051086


In [ ]:
df.groupby("major_category")["major"].unique()

,major
major_category,
STEM,"[biology, electrical_engineering, mechanical_e..."
arts,[fine_arts]
business,"[marketing, business, economics, finance, busi..."
education,[education]
health,[nursing]
humanities,[humanities]
social_science,"[social_work, communications, criminal_justice..."


Sprawdzenie struktury kategorii wykazało, że część z nich obejmuje tylko jeden kierunek, m.in. arts, humanities, education i health. W ich przypadku analiza na poziomie kierunku nie stanowi dodatkowego potwierdzenia wyników kategorii. Większą wartość analityczną ma porównanie kierunków wewnątrz bardziej zróżnicowanych kategorii, takich jak social_science, business i STEM.